[TG-8293] Constant propagation for empty strings and string concatenation [blocks: #4941]

danpoe opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

danpoe commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,141 @@ void symex_assignt::assign_from_struct(
+    {});
+
+  symbol_table_baset &st_mutable =
+    const_cast<symbol_table_baset &>(ns.get_symbol_table());
```
:skull:

romainbrenguier replied:
:no_entry_sign:

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -14,9 +14,17 @@ Author: Daniel Kroening, kroening@kroening.com
+#include <util/mathematical_expr.h>
+#include <util/string_expr.h>
+
+#include <iostream>
```
probably not necessary

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+  const ssa_exprt &char_array,
+  const array_exprt &new_char_array)
+{
+  // assign length of string
```
:pick: I find a comment like // length = new_length would be clearer

romainbrenguier commented on src/goto-symex/symex_assign.h:
```diff
@@ -109,6 +115,15 @@ class symex_assignt
+  bool constant_propagate_empty_string(
+    const function_application_exprt &f_l1);
+
+  void assign_constants(
```
:no_entry_sign: documentation is missing for these two functions

romainbrenguier commented on src/goto-symex/symex_assign.h:
```diff
@@ -35,7 +36,9 @@ class symex_assignt
-      target(target)
+      target(target),
+      char_type(16),
+      length_type(32)
```
these fields shouldn't be necessary and this is not language independent like symex should be.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -378,7 +472,41 @@ void symex_assignt::assign_non_struct_symbol(
+      INVARIANT(guard.empty(), "assignment does not have a guard");
+      INVARIANT(
+        rhs.id() == assignment.rhs.id(),
+        "rhs has same kind after renaming and field sensitivity processing");
```
I'm not sure of the reason for these invariant, maybe they are just needed for `constant_propagate_empty_string`, in which case they should come later.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -378,7 +472,41 @@ void symex_assignt::assign_non_struct_symbol(
+        {
+          ret = constant_propagate_empty_string(f_l1);
+        }
+
```
too many blank lines

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -378,7 +472,41 @@ void symex_assignt::assign_non_struct_symbol(
+        const irep_idt &func_id =
+          to_symbol_expr(f_l1.function()).get_identifier();
+
+        int ret = false;
```
:no_entry_sign: types do not match

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -378,7 +472,41 @@ void symex_assignt::assign_non_struct_symbol(
+        to_function_application_expr(rhs);
+
+      const function_application_exprt &f_l2 =
+        to_function_application_expr(assignment.rhs);
```
I don't see where this is used

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -378,7 +472,41 @@ void symex_assignt::assign_non_struct_symbol(
+      if(f_l1.function().id() == ID_symbol)
+      {
+        const irep_idt &func_id =
+          to_symbol_expr(f_l1.function()).get_identifier();
```
can be inlined in the if

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -378,7 +472,41 @@ void symex_assignt::assign_non_struct_symbol(
+
+        int ret = false;
+
+        if(func_id == ID_cprover_string_empty_string_func)
```
the expression for ret could be simplified into a conjunction

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+
+  array_typet new_char_array_type(
+    char_type,
+    new_length);
```
:pick: brace `{ }` initialization is preferred. :baby_chick:

romainbrenguier replied:
const

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+  aux_symbol.is_static_lifetime = true;
+  aux_symbol.is_file_local = false;
+  aux_symbol.is_thread_local = false;
+
```
:pick: unnecessary blank line

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+
+  aux_symbol.value = new_char_array;
+
+  exprt arr =
```
const ?

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+    new_char_array,
+    {});
+
+  address_of_exprt aoe(
```
const

romainbrenguier replied:
:baby_chick:

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+    new_char_array,
+    {});
+
+  address_of_exprt aoe(
```
give a better name than `aoe`

romainbrenguier replied:
alternatively inline this expression in the next call

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+      aux_symbol.symbol_expr(),
+      from_integer(0, length_type)));
+
+  assign_symbol(
```
:pick: add a comment like // char_array = &aux_symbol[0]

romainbrenguier replied:
:strawberry:

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+    char_type,
+    new_length);
+
+  symbolt &aux_symbol = get_fresh_aux_symbol(
```
:pick: give a better name than `aux_symbol`, one that would make clear that it is a char array

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+  const array_exprt &new_char_array)
+{
+  // assign length of string
+  assign_symbol(
```
since you know the rhs is not a `struct` you can call directly `assign_non_struct_symbol` :strawberry:

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+        ns)
+      .get();
+
+  assign_symbol(
```
:strawberry:

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+{
+  const constant_exprt length = from_integer(0, length_type);
+
+  array_typet char_array_type(
```
length_type and char_type should be deduced from the type of the arguments of f_l1

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+}
+
+bool symex_assignt::constant_propagate_empty_string(
+  const function_application_exprt &f_l1)
```
:pick: could we use a clearer name, like `function_application_l1`?

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+
+  INVARIANT(
+    f_l1.arguments().size() == 2,
+    "empty string primitive does not take arguments");
```
not very clear because it asserts there are 2 arguments, but the comment says there are 0

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
   }
 }
 
+void symex_assignt::assign_constants(
```
should be made clear in the name that it is for string constants

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -350,6 +358,92 @@ void symex_assignt::assign_from_struct(
+    f_l1.arguments().size() == 2,
+    "empty string primitive does not take arguments");
+
+  array_exprt char_array({}, char_array_type);
```
:baby_chick:

romainbrenguier reviewed (changes requested):
Review for the first commit

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -444,6 +444,55 @@ bool symex_assignt::constant_propagate_empty_string(
+  const function_application_exprt &f_l1,
+  const function_application_exprt &f_l2)
+{
+  const refined_string_exprt &s1 = to_string_expr(f_l2.arguments().at(2));
```
used only once, so can be inlined below. There are a lot of variables representing basically the same thing in this function (s1, s1_data_opt, s1_data) so it's bette to try to minimize that :lemon:

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -444,6 +444,55 @@ bool symex_assignt::constant_propagate_empty_string(
+
+  const array_exprt &s1_data = s1_data_opt->get();
+
+  const refined_string_exprt &s2 = to_string_expr(f_l2.arguments().at(3));
```
:lemon:

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -444,6 +444,55 @@ bool symex_assignt::constant_propagate_empty_string(
+  const constant_exprt new_char_array_length =
+    from_integer(new_size, length_type);
+
+  array_typet new_char_array_type(
```
:strawberry:

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -444,6 +444,55 @@ bool symex_assignt::constant_propagate_empty_string(
+  operands.insert(
+    operands.end(),
+    s2_data.operands().begin(),
+    s2_data.operands().end());
```
Ideally we should avoid duplication with the code from `solver/strings/string_concatenation_builtin_function.h`, concatenation is pretty simple and unlikely to change, but for functions like format we definitely need to share the code

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -444,6 +444,55 @@ bool symex_assignt::constant_propagate_empty_string(
+
+  array_exprt new_char_array(
+    std::move(operands),
+    new_char_array_type);
```
:strawberry:

romainbrenguier commented on src/util/simplify_expr.h:
```diff
@@ -27,4 +31,9 @@ bool simplify(
 exprt simplify_expr(exprt src, const namespacet &ns);
 
+optionalt<std::reference_wrapper<const array_exprt>>
+try_get_string_data_array(
```
:no_entry_sign:  needs documentation

romainbrenguier commented on src/util/simplify_expr.h:
```diff
@@ -27,4 +31,9 @@ bool simplify(
 // this is the preferred interface
 exprt simplify_expr(exprt src, const namespacet &ns);
 
+optionalt<std::reference_wrapper<const array_exprt>>
```
:pick: optional references can be replaced by pointers

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 19f8245930aa1cf3240fdee9344dc58878559ca3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119510875

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 686dd8ae485b112206f693a5365d0086dd02524f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120002129
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d0b5245dc260be952c0880e18ec547d8a2e11afc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120004102

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0db7b6ecac31904256dd030c0345be226b9c875d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120051213

smowton commented on jbmc/regression/jbmc-strings/StringBuilderAppendConstantEvaluation1/Main.java:
```diff
@@ -0,0 +1,9 @@
+    sb.append("xyz");
+    String s = sb.toString();
+    assert s.length() == 6;
+    assert s.startsWith("abcxyz");
```
Surely the tests should check `equals`; whether that decomposes into `startsWith` and `length` is the model's business

danpoe replied:
This test doesn't use the models. I've not included them for tests where they are not needed.

smowton commented on jbmc/regression/jbmc-strings/StringConcatenationConstantEvaluation3/Main.java:
```diff
@@ -0,0 +1,8 @@
+public class Main {
+  public void test(String s1) {
+    String s2 = "abc";
+    assert (s1 + s2).startsWith("abc");
```
Comment either here or in the `.desc` files that we're testing that constant propagation does not happen in these cases

smowton commented on jbmc/regression/jbmc-strings/StringConcatenationConstantEvaluation3/Main.java:
```diff
@@ -0,0 +1,8 @@
+  public void test(String s1) {
+    String s2 = "abc";
+    assert (s1 + s2).startsWith("abc");
+    assert ("" + s1).startsWith("abc");
```
The way these are sequenced means there's no consistent value of `s1` that will get through the first two assertions, so the last one is actually dead code -- suggest putting these in separate functions and using `--function` instead of `--property`.

smowton commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -80,8 +86,204 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+        return constant_propagate_string_concat(
+          state, symex_assign, f_l1, f_l2);
+      }
+      else if(func_id == ID_cprover_string_empty_string_func)
```
Surprising this is handled in `_with_side_effects`

danpoe replied:
The assignment of the empty string is a side effect of the whole assignment `lhs = f(length, data)` with `f` being a function application expression with ID `ID_cprover_string_empty_string_func` and `length` and `data` being the output parameters to which the empty string length and data array pointer are assigned to.

smowton replied:
and again `lhs` is not used -- this is purely a constraint on `length` and `data`, right? In which case this should be operating over expressions, not assignments

allredj commented on jbmc/regression/jbmc-strings/StringBuilderAppendConstantEvaluation1/Main.java:
```diff
@@ -0,0 +1,9 @@
+    sb.append("xyz");
+    String s = sb.toString();
+    assert s.length() == 6;
+    assert s.startsWith("abcxyz");
```
:no_entry_sign: I think it's not sufficient to have tests for the successful case. We should have tests that check some assertion is violated.
Otherwise we're only checking for under-approximation, and assertions are true when unreachable.
See https://github.com/danpoe/cbmc/tree/feature/simplify-expression-strings/jbmc/regression/jbmc-strings/CompareToConstantEvaluation

danpoe replied:
Test added by @yumibagge

allredj commented on src/goto-symex/goto_symex.h:
```diff
@@ -528,6 +530,60 @@ class goto_symext
+  /// \param state: goto symex state
+  /// \param symex_assign: object handling symbol assignments
+  /// \param lhs: lhs of the assignment
+  /// \param rhs: rhs of the assignment
```
Please document return value :hamster:

allredj commented on src/goto-symex/goto_symex.h:
```diff
@@ -528,6 +530,60 @@ class goto_symext
+  /// \param f_l1: application of function ID_cprover_string_concat_func with
+  ///   l1 renaming applied
+  /// \param f_l2: same as above but with l2 renaming applied
+  bool constant_propagate_string_concat(
```
Please document return value :hamster:

allredj commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -162,7 +162,10 @@ void array_poolt::insert(
+  if(length_of_array.find(array_expr) == length_of_array.end())
+  {
+    attempt_assign_length_from_type(array_expr, length_of_array, fresh_symbol);
+  }
```
Some of these changes look like they should be in separate commits. :lock:

allredj commented on src/util/simplify_expr.cpp:
```diff
@@ -1673,10 +1668,8 @@ optionalt<std::string> simplify_exprt::expr2bits(
-    const refined_string_exprt &s,
-    const namespacet &ns)
+optionalt<std::reference_wrapper<const array_exprt>>
+try_get_string_data_array(const refined_string_exprt &s, const namespacet &ns)
```
:lock:

allredj commented on src/util/simplify_expr.cpp:
```diff
@@ -64,11 +64,6 @@ struct simplify_expr_cachet
-try_get_string_data_array(
-  const refined_string_exprt &s,
-  const namespacet &ns);
-
```
:lock:

smowton commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -80,8 +86,204 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+
+  aux_symbol.value = new_char_array;
+
+  exprt arr = state.rename(aux_symbol.symbol_expr(), ns).get();
```
`aux_symbol_expr`?

smowton commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -80,8 +86,204 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  return false;
+}
+
+void goto_symext::assign_string_constant(
```
Comment on what this actually does -- generate assignments `length = new_length`, `char aux[new_length]; aux = new_char_array; char_array = &aux[0]; dummy = associate_array_to_pointer(new_char_array, &aux[0])`

danpoe replied:
Expanded the documentation in `goto_symex.h`.

smowton commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -80,8 +86,204 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
   }
 }
+
+bool goto_symext::constant_propagate_assignment_with_side_effects(
```
Suggest a better term than  `assignment_with_side_effects` might be `constraint`? It looks like this is constraining its two operands to have a particular relationship, rather than acting as the pure function the `ID_function_application` suggests. @allredj thoughts?

danpoe replied:
The `lhs` argument is currently not used (it's not needed for the string operations currently handled but might be needed in the future). The `rhs` is a `function_application_exprt` with the first two arguments being output parameters (so it's not a pure a function, a bit unfortunate I know). So overall here we handle the potential side effects of `lhs = rhs` since the evaluation of the `rhs` has side effects.

smowton replied:
It's a bit weird that these only apply at the root of an assignment though. How about instead recursing over the LHS and RHS with something like `propagate_expression_side_effects`, and if it wanted to also propagate a return value it would just replace the expression with its result and let that be propagated as an ordinary assignment?

danpoe replied:
We could of course recurse into the operands. Since we currently don't need that though, can we leave this for a future PR?

peterschrammel replied:
>  but might be needed in the future

@danpoe, are there any operations where this is actually used?

danpoe replied:
No, and I think that's ok.

smowton commented on src/goto-symex/goto_symex.h:
```diff
@@ -528,6 +530,60 @@ class goto_symext
+    symex_assignt &symex_assign,
+    const function_application_exprt &f_l1);
+
+  /// Attempt to constant propagate string concatenation
```
return value? (here and elsewhere)

smowton commented on src/goto-symex/symex_assign.h:
```diff
@@ -17,6 +17,7 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
 
 class byte_extract_exprt;
 class expr_skeletont;
+class function_application_exprt;
```
This looks like it belongs in the cpp file, since we haven't changed anything here

smowton reviewed (commented):
I note you're generating a fresh constant char array every time this happens, which particularly for the empty string function is likely to generate many duplicates. How about naming the constant arrays after their constants, as in the Java frontend (https://github.com/diffblue/cbmc/blob/develop/jbmc/src/java_bytecode/java_string_literals.cpp#L59), so you can avoid recreating the same array again and again?

allredj commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -80,8 +86,204 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  if(!symex_config.simplify_opt)
+  {
+    return false;
+  }
```
:bulb: Somehow I would prefer this check to be external to the function. If you want to keep it in the behaviour should be documented in the function description.

allredj commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -80,8 +86,204 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+    return false;
+  }
+
+  if(state.threads.size() > 1)
```
Needs some documentation (preferably in the function documentation).

allredj commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -80,8 +86,204 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  {
+    exprt rhs_l2 = state.rename(rhs, ns).get();
+
+    const function_application_exprt &f_l1 = to_function_application_expr(rhs);
```
Move this assignment below the `f_l2` one (to have it closer to its usage).

allredj commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -80,8 +86,204 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+
+  if(rhs.id() == ID_function_application)
+  {
+    exprt rhs_l2 = state.rename(rhs, ns).get();
```
@danpoe I don't really understand why we do an L2 renaming here. Can we look at this together please?

danpoe replied:
Ah, I initially did this since `l2` renaming also replaces `l1` symbols with `l2` expressions stored in the propagation map apparently. But I think I can retrieve the expressions from the propagation map directly instead of doing the renaming.

allredj replied:
@danpoe, so are you doing this?

danpoe replied:
Ah yes this code can now be removed.

allredj reviewed (commented):
Looks generally good to me.
One thing to addess is that we're loosing test coverage for `add_axioms_for_empty_string()` in src/solvers/strings/string_constraint_generator_constants.cpp and a couple of lines in src/solvers/strings/string_concatenation_builtin_function.cpp.
I suppose some of these tests now get caught into constant propagation (which is good sign) but the tests need to be adapted so they don't.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 085828faedd37018456a91997f8fce45d8c2fffc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120752999
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

danpoe committed:
Move escape_non_alnum() to util/string_utils.{h,cpp}

The function is generally useful for obtaining strings that contain only
alphanumeric characters.

yumibagge committed:
Add regression test that uses models-library

yumibagge committed:
Add regression test to check verification failure

yumibagge committed:
Add regression test for 2byte char

yumibagge committed:
Add regression test for non printable character

danpoe committed:
Check that `array_expr` is not in `length_of_array` map in array_pool.cpp

Previously the string solver returned an error when constant propagation of
strings was used as `array_poolt::insert()` might be called for array
expressions for which there is already a mapping in the `length_of_array` map.

danpoe committed:
Make try_get_string_data_array() non-static

This is so the function can be called outside of simplify_expr.cpp.

danpoe committed:
Implement constant propagation of string concatenation

This implements constant propagation of ID_cprover_string_concat_func and
ID_cprover_string_empty_string_func which are the primitives used by
StringBuilder.append(), which in turn is used to implement String concatenation
using +.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a9af62f634e3be3c700a8718281aeccaca2138fa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120981015

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bab613afca0f46cae9e5f8f37e9c8a21042e6d46).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120981224

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -80,8 +89,257 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+
+  array_typet new_char_array_type = new_char_array.type();
+
+  new_char_array_type.set(ID_C_constant, true);
```
The blank line is distracting

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -80,8 +89,257 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+
+    new_aux_symbol.value = new_char_array;
+
+    exprt arr = state.rename(new_aux_symbol.symbol_expr(), ns).get();
```
const?

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -80,8 +89,257 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+      "symbol shall have value derived from char array content");
+  }
+
+  address_of_exprt aoe(
```
avoid abbreviations like `aoe`

danpoe committed:
Change interface of try_get_string_data_array()

Take the content field of a refined string expression instead of taking the
whole refined string expression. This will allow it to be used from the string
mutators constant propagation code in goto_symex.cpp.

allredj commented on jbmc/regression/jbmc-strings/ConstantEvaluationStringConcatenation05-VerificationFailure/test.desc:
```diff
@@ -0,0 +1,8 @@
+--function Main.test --property "java::Main.test:()V.assertion.1" --property "java::Main.test:()V.assertion.2"
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
```
Does it make sense to check for two properties at the same time?
I think this only checks that at least one of them fails.

allredj commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,277 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  const constant_exprt &new_length,
+  const ssa_exprt &char_array,
+  const array_exprt &new_char_array)
+{
```
:pick: It looks like this function could do with some refactoring. It's a bit too long to be readable. Extracting parts into named functions would make it more understandable.

danpoe replied:
Ok, added to helper functions being called by this function.

allredj commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,277 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  return try_get_string_data_array(s_pointer_opt->get(), ns);
+}
+
+std::string goto_symext::get_alnum_string(const array_exprt &char_array)
```
Documentation would be useful.

danpoe replied:
Documented in the header file.

allredj replied:
:+1:  Indeed. I don't know how I missed it...

allredj commented on jbmc/regression/jbmc-strings/ConstantEvaluationStringConcatenation07-NonPrintableChar/test.desc:
```diff
@@ -0,0 +1,9 @@
+^Generated [0-9]+ VCC\(s\), 0 remaining after simplification$
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
```
Sorry for being a bore, but regression tests that check for success only verify that cbmc doesn't do anything. They should always be paired with a test checking for failure.

allredj replied:
After discussion with @danpoe, it seems to be unnecessary to "assertion failure" tests for each case, as not many bad things can happen.
We do want to check that we're not completely discarding the trace, so a few tests for assertion failure makes sense. The tests currently included now look sufficient to me.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: cd8a0a8ca7a254abea6ef812e4ec938d8234511e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/121103986

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: cdf4102b9257f1edb78b7e3ad1a66420f9b82a5c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/121115810

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: bbbe08c3b88a55755f8471bd1079ca9fac3e148c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/121121929
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 72624884ae1cce9e66831abaed86b4f175df22d2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/121134021

romainbrenguier commented on src/util/simplify_expr.h:
```diff
@@ -31,19 +31,19 @@ bool simplify(
 exprt simplify_expr(exprt src, const namespacet &ns);
 
-/// Get char sequence from refined string expression
+/// Get char sequence from content() field of a refined string expression
```
:pick: `content()` -> `content`

romainbrenguier commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -162,7 +162,10 @@ void array_poolt::insert(
     it_bool.second, "should not associate two arrays to the same pointer");
 
-  attempt_assign_length_from_type(array_expr, length_of_array, fresh_symbol);
+  if(length_of_array.find(array_expr) == length_of_array.end())
```
:bulb: in test-gen repository we have a `has` function which could be nice to use in cbmc too

romainbrenguier replied:
The commit message seems to say the opposite of the change since the function is called only when there is no mapping.

romainbrenguier commented on jbmc/regression/jbmc-strings/StringConcatenationConstantEvaluation3/test1.desc:
```diff
@@ -0,0 +1,10 @@
+--function Main.test1 --property "java::Main.test1:(Ljava/lang/String;)V.assertion.1" --show-vcc
+^Generated [0-9]+ VCC\(s\), 1 remaining after simplification$
+^EXIT=0$
+^SIGNAL=0$
```
Check also for VERIFICATION FAILED? :bird:

romainbrenguier commented on jbmc/regression/jbmc-strings/StringConcatenationConstantEvaluation3/test2.desc:
```diff
@@ -0,0 +1,10 @@
+--function Main.test2 --property "java::Main.test2:(Ljava/lang/String;)V.assertion.1" --show-vcc
+^Generated [0-9]+ VCC\(s\), 1 remaining after simplification$
+^EXIT=0$
+^SIGNAL=0$
```
:bird:

romainbrenguier commented on jbmc/regression/jbmc-strings/StringConcatenationConstantEvaluation3/test3.desc:
```diff
@@ -0,0 +1,10 @@
+--function Main.test3 --property "java::Main.test3:(Ljava/lang/String;)V.assertion.1" --show-vcc
+^Generated [0-9]+ VCC\(s\), 1 remaining after simplification$
+^EXIT=0$
+^SIGNAL=0$
```
:bird:

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+
+  DATA_INVARIANT(
+    f_l1.arguments().size() == 2,
+    "empty string primitive takes two output arguments");
```
this should be a precondition of the function

danpoe replied:
A precondition of the method is that `f_l1` is a `function_application_exprt` with ID `ID_cprover_string_empty_string_func`. So at this point it feels more appropriate to phrase the condition on the number of arguments as a `DATA_INVARIANT`. Plus it documents the fact that we have two output parameters.

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+
+  INVARIANT(
+    aux_symbol.value == new_char_array,
+    "symbol shall have value derived from char array content");
```
:warning: This is relying on symbols ending with `_constant_char_array` only being used to represent string constant by this function.

danpoe replied:
True. This uses `state.symbol_table` which contains only the symbols minted during symbolic execution so the naming is within our control.

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  new_aux_symbol.is_file_local = false;
+  new_aux_symbol.is_thread_local = false;
+
+  new_aux_symbol.value = new_char_array;
```
:pick: this blank line can be removed

danpoe replied:
Intentional for better readability.

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  const symbolt &function_symbol =
+    ns.lookup(ID_cprover_associate_array_to_pointer_func);
+
+  function_application_exprt fae(
```
:pick: avoid abbreviations like fae :fox_face:

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+
+std::string goto_symext::get_alnum_string(const array_exprt &char_array)
+{
+  const auto &ibv_type =
```
:fox_face:

danpoe replied:
Abbreviation is fine imo

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+
+  static_assert(CHAR_BIT == 8, "bitwidth of char assumed to be 8");
+
+  const std::size_t n = n_bits / 8;
```
Better name than `n`?

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+
+  const constant_exprt length = from_integer(0, length_type);
+
+  array_typet char_array_type(char_type, length);
```
`const` or inline in the `char_array` declaration

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+    f_l1.arguments().size() == 2,
+    "empty string primitive takes two output arguments");
+
+  array_exprt char_array({}, char_array_type);
```
const or inline below :bee:

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  const auto &length_type = f_type.domain().at(0);
+  const auto &char_type = to_pointer_type(f_type.domain().at(1)).subtype();
+
+  const refined_string_exprt &s1 = to_string_expr(f_l1.arguments().at(2));
```
blank line above can be removed to group abbreviations together

danpoe replied:
Prefer with blank line for readability.

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  const refined_string_exprt &s1 = to_string_expr(f_l1.arguments().at(2));
+  const auto s1_data_opt = try_evaluate_constant_string(state, s1.content());
+
+  if(!s1_data_opt)
```
:pick: I find it more logical to but the blank line before `s1_data_opt` definition rather than after :lollipop:

danpoe replied:
I prefer the current version, and to always have a blank line before an `if`.

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  const refined_string_exprt &s2 = to_string_expr(f_l1.arguments().at(3));
+  const auto s2_data_opt = try_evaluate_constant_string(state, s2.content());
+
+  if(!s2_data_opt)
```
:lollipop:

danpoe replied:
Prefer current version.

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  const constant_exprt new_char_array_length =
+    from_integer(new_size, length_type);
+
+  array_typet new_char_array_type(char_type, new_char_array_length);
```
:bee:

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  operands.insert(
+    operands.end(), s2_data.operands().begin(), s2_data.operands().end());
+
+  array_exprt new_char_array(std::move(operands), new_char_array_type);
```
:bee:

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  return try_get_string_data_array(s_pointer_opt->get(), ns);
+}
+
+std::string goto_symext::get_alnum_string(const array_exprt &char_array)
```
:no_entry_sign: This shouldn't be a member method of `goto_symext`, it's not using any of its fields

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,285 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  return new_aux_symbol;
+}
+
+void goto_symext::associate_array_to_pointer(
```
It doesn't look like this needs to be a member function of `goto_symext` (it only uses `ns` which can be deduced from the symbol table)

danpoe replied:
We need the namespace here as it contains two symbol tables (those from the goto program and those minted during symex).

romainbrenguier replied:
`ns` can be given as an argument

romainbrenguier reviewed (dismissed):
The `goto_symext` class is already too big, I'd prefer we try to not add to much to it. In particular functions which do not touch its private fields should not be members of `goto_symext`.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 16becf6a196d4abe68db101d456e06918e6bd95d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/121290733

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 71d11d56f4ee3e659aea31740e6f1380b51deed3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/121745220

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,294 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  const auto &ibv_type =
+    to_integer_bitvector_type(to_array_type(char_array.type()).subtype());
+
+  const std::size_t n_bits = ibv_type.get_width();
```
:pick: `ibv_type` is used only once, it can be inlined here since the variable name doesn't give any information.

danpoe replied:
Does provide information, and is more readable as is imo

romainbrenguier replied:
I wouldn't know what `ibv` meant if it wasn't followed by `to_integer_bitvector_type`, so at least to me it gives no information

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,294 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+    ns,
+    state.symbol_table);
+
+  ssa_exprt ssa_expr(return_symbol.symbol_expr());
```
const

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,294 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  function_application_exprt array_to_pointer_app(
+    function_symbol.symbol_expr(), {new_char_array, string_data});
+
+  symbolt &return_symbol = get_fresh_aux_symbol(
```
const

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,294 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  const symbolt &function_symbol =
+    ns.lookup(ID_cprover_associate_array_to_pointer_func);
+
+  function_application_exprt array_to_pointer_app(
```
const and `{` `}` initialization is prefered

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,294 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+    aux_symbol.value == new_char_array,
+    "symbol shall have value derived from char array content");
+
+  address_of_exprt string_data(
```
const

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,294 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  return false;
+}
+
+void goto_symext::assign_string_constant(
```
I don't think this function belongs in the `goto_symext` class, it doesn't seem to use any of its field (apart from ns which could be given as argument). :octopus:

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,294 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  }
+}
+
+const symbolt &goto_symext::get_new_string_data_symbol(
```
:octopus:

romainbrenguier commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,294 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  return try_get_string_data_array(s_pointer_opt->get(), ns);
+}
+
+void goto_symext::constant_propagate_empty_string(
```
:octopus:

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: e78d73ad4b454aa09daf107717962b4d1c69aa4c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122123110
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

romainbrenguier commented on src/goto-symex/goto_symex.h:
```diff
@@ -528,6 +530,103 @@ class goto_symext
+  /// \param symex_assign: object handling symbol assignments
+  /// \param f_l1: application of function ID_cprover_string_empty_string_func
+  ///   with l1 renaming applied
+  void constant_propagate_empty_string(
```
Sorry, I forgot to ask for unit tests for this function and the one below (ideally also the others that are added to `goto_symext`)

danpoe commented:
Regarding @romainbrenguier's comment above, we have agreed that adding unit tests for the added methods in `goto_symext` is not strictly necessary as the feature is well-tested by regression tests already. The "don't like the changes" refers to some of the added methods being members of `goto_symext` as opposed to being static functions (in the same file).

peterschrammel commented on jbmc/regression/jbmc-strings/ConstantEvaluationStringConcatenation07-NonPrintableChar/Main.java:
```diff
@@ -0,0 +1,9 @@
+    String s2 = "\\";
+    String s3 = s1 + s2;
+    assert s3.length() == 2;
+    assert s3.startsWith("\t");
```
Maybe add `assert s3.endsWith("\\");`

danpoe replied:
Ok

danpoe replied:
`s1.equals(s2)` is implemented in the models as `s1.length() == s2.length() && s1.startsWith(s2)` so here we're checking that the strings are equal. To keep the tests as simple as possible we're not using the models library. I don't think it's necessary to use `endsWith()` here as this is a test for concatenation, `endsWith()` is tested elsewhere.

peterschrammel replied:
Then it should better be `assert s3.startsWith("\t\\");` if you want to emulate equals?

peterschrammel commented on jbmc/regression/jbmc-strings/StringBuilderDefaultConstructorConstantEvaluation/Main.java:
```diff
@@ -0,0 +1,8 @@
+    StringBuilder sb = new StringBuilder();
+    String s = sb.toString();
+    assert s.isEmpty();
+    assert s.startsWith("");
```
Isn't that always `true`?

danpoe replied:
Yes it should be.

peterschrammel commented on jbmc/regression/jbmc-strings/StringBuilderAppendConstantEvaluation1/test.desc:
```diff
@@ -0,0 +1,9 @@
+CORE
+Main.class
+--function Main.test --property "java::Main.test:()V.assertion.1" --property "java::Main.test:()V.assertion.2"
+^Generated [0-9]+ VCC\(s\), 0 remaining after simplification$
```
Shouldn't this be exactly `1 VCC` if you use `--property`?

danpoe replied:
`--property` is passed in twice here, so yes I think the value should be 2. I'll change it to use the precise value.

danpoe replied:
So it turns out that `jbmc` translates those assertions to something like:
```
if(length_return_value != 6)
{
  assert(false);
}
```
So if `length_return_value` is 6, we don't symex the assertion and thus get 0 VCCs before simplification. This is probably not stable behaviour though so I'll leave it as is. The important thing is that after simplification we have 0 remaining VCCs.

peterschrammel commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -87,8 +96,294 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+}
+
+optionalt<std::reference_wrapper<const array_exprt>>
+goto_symext::try_evaluate_constant_string(
```
🐙

danpoe replied:
I prefer those methods to live in `goto_symext` for now as opposed to being static. The plan is to eventually create a `goto_symex_constant_propagationt` member of `goto_symext` which handles the constant propagation.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4fecf4ec42b96ee26398ef99e5bf002f8d5e25eb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122424687

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d9f75eb459feee99ee53921418b7ed94f2ad5b2f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122439299
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
